microcosm_pubsub/decorators/publish_decorators.py

```diff
@@ -0,0 +1,138 @@
+    :media_type      - the message media type. If is not specified, will try to create: created(component.ns.subject)
+    :default_uri     - boolean, should autogenerate uri query string arg. If set to true,
+                       Uses operation argument, component.identifier_key and function_result.id to generate uri
+    :uri_string_args - Allow to specify query args to use with the default_uri instead of the default identifier_key.
+                       Passed as a dictionary of {arg_name: lambda function_result: action}
+                       For example: uri_string_args={
+                           company_event_id: lambda company_event: company_event.id
+                           min_clock: lambda company_event: company_event.clock
+                        }
+    :message_params  - Allow to specify additional message args to use with the default_uri.
+                       Passed as a dictionary of {arg_name: lambda function_result: action}
+                       For example: uri_string_args={ company_id: lambda company_event: company_event.company_id }
+                       If default_uri is set to false, allows to pass uri argument for non standard uris
+                       For example: uri_string_args={ uri: lambda resource: resource["_links"]["self"]["href"] }
+                       It is recommended to set both media_type and message_params toghether.
+    :ns              - If ns is set to true or if media_type is unspecified,
+                       allows to specify any ns. Useful if component.ns is not set.
+    :identifier_key  - If default_uri is set to true, allows to specify any identifier_key.
+                       The value will still be model.id
+                       Useful if component.identifier_key is not set or when setting the ns arg.
+    :operation       - microcosm_flask.operations.Operation to use for autogenerating the default_uri.
+                       Default is Operation.Retrieve.
+                       It is recommended to set both operation and uri_string_args toghether.
+
+    """
+    def decorator(func):
+        if default_uri and message_params and "uri" in message_params:
+            raise ValidationError("Cannot pass both uri extractor and while default_uri set to true")
+        operation_ = operation or Operation.Retrieve
+
+        @wraps(func)
+        def decorate(component, *func_args, **func_kwargs):
+            sns_producer = getattr(component, producer_key, None)
+            if sns_producer is None:
+                sns_producer = getattr(component.graph, producer_key)
+            media_type_to_publish = media_type or _create_default_media_type(component, **kwargs)
+
+            extractors = dict(message_params) if message_params else dict()
+            if default_uri:
+                uri = _create_default_uri_extractor(component, operation_, uri_string_args, **kwargs)
```

Comment:
I think you want to have a `uri_factory` or `uri_extractor` be injected into this function and have `microcosm-flask` define an implementation that is decoupled from publishing.